// ng2-redux-router/lib/es6/router.js
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/map';
import { Injectable, ApplicationRef } from '@angular/core';
import { Location } from '@angular/common';
import { Router, NavigationEnd } from '@angular/router';
import { NgRedux } from 'ng2-redux';
import { UPDATE_LOCATION } from './actions';
export class NgReduxRouter {
    constructor(router, ngRedux, applicationRef, location) {
        this.router = router;
        this.ngRedux = ngRedux;
        this.applicationRef = applicationRef;
        this.location = location;
        this.initialized = false;
        this.selectLocationFromState = (state) => state.router;
    }
    /**
     * Destroys the bindings between ng2-redux and @angular/router.
     * This method unsubscribes from both ng2-redux and @angular router, in case
     * your app needs to tear down the bindings without destroying Angular or Redux
     * at the same time.
     */
    destroy() {
        if (this.urlStateSubscription) {
            this.urlStateSubscription.unsubscribe();
        }
        if (this.reduxSubscription) {
            this.reduxSubscription.unsubscribe();
        }
        this.initialized = false;
    }
    /**
     * Initialize the bindings between ng2-redux and @angular/router
     *
     * This should only be called once for the lifetime of your app, for
     * example in the constructor of your root component.
     *
     *
     * @param {(state: any) => string} selectLocationFromState Optional: If your
     * router state is in a custom location, supply this argument to tell the
     * bindings where to find the router location in the state.
     * @param {Observable<string>} urlState$ Optional: If you have a custom setup
     * when listening to router changes, or use a different router than @angular/router
     * you can supply this argument as an Observable of the current url state.
     */
    initialize(selectLocationFromState = (state) => state.router, urlState$ = undefined) {
        if (this.initialized) {
            throw new Error('ng2-redux-router already initialized! If you meant to re-initialize, call destroy first.');
        }
        this.selectLocationFromState = selectLocationFromState;
        this.urlState = urlState$ || this.getDefaultUrlStateObservable();
        this.listenToRouterChanges();
        this.listenToReduxChanges();
        this.initialized = true;
    }
    getDefaultUrlStateObservable() {
        return this.router.events
            .filter(event => event instanceof NavigationEnd)
            .map(event => this.location.path())
            .distinctUntilChanged();
    }
    getLocationFromStore(useInitial = false) {
        return this.selectLocationFromState(this.ngRedux.getState()) ||
            (useInitial ? this.initialLocation : '');
    }
    listenToRouterChanges() {
        const handleLocationChange = (location) => {
            if (this.currentLocation === location) {
                // Dont dispatch changes if we haven't changed location.
                return;
            }
            this.currentLocation = location;
            if (this.initialLocation === undefined) {
                this.initialLocation = location;
                // Fetch initial location from store and make sure
                // we dont dispath an event if the current url equals
                // the initial url.
                let locationFromStore = this.getLocationFromStore();
                if (locationFromStore === this.currentLocation) {
                    return;
                }
            }
            this.ngRedux.dispatch({
                type: UPDATE_LOCATION,
                payload: location
            });
        };
        this.urlStateSubscription = this.urlState.subscribe(handleLocationChange);
    }
    listenToReduxChanges() {
        const handleLocationChange = (location) => {
            if (this.initialLocation === undefined) {
                // Wait for router to set initial location.
                return;
            }
            let locationInStore = this.getLocationFromStore(true);
            if (this.currentLocation === locationInStore) {
                // Dont change router location if its equal to the one in the store.
                return;
            }
            this.currentLocation = location;
            this.router.navigateByUrl(location);
        };
        this.reduxSubscription = this.ngRedux
            .select(state => this.selectLocationFromState(state))
            .distinctUntilChanged()
            .subscribe(handleLocationChange);
    }
}
NgReduxRouter.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NgReduxRouter.ctorParameters = [
    { type: Router, },
    { type: NgRedux, },
    { type: ApplicationRef, },
    { type: Location, },
];
